libc-bottom-half/cloudlibc/src/libc/unistd/write.c

```diff
@@ -2,11 +2,131 @@
 //
 // SPDX-License-Identifier: BSD-2-Clause
 
+#ifdef __wasilibc_use_wasip2
+#include <wasi/wasip2.h>
+#include <wasi/descriptor_table.h>
+#include <common/errors.h>
+#include <time.h>
+#else
 #include <wasi/api.h>
+#endif
 #include <errno.h>
 #include <unistd.h>
 
 ssize_t write(int fildes, const void *buf, size_t nbyte) {
+#ifdef __wasilibc_use_wasip2
+  streams_borrow_output_stream_t output_stream;
+  filesystem_own_input_stream_t input_stream;
+  bool ok = false;
+  bool create_new_stream = false;
+  filesystem_error_code_t error_code;
+  descriptor_table_entry_t* entry = 0;
+
+  // Check for stdin/stdout/stderr
+  if (fildes == 1)
+    output_stream = streams_borrow_output_stream(stdout_get_stdout());
+  else if (fildes == 2)
+    output_stream = streams_borrow_output_stream(stderr_get_stderr());
+  else {
+    // Translate the file descriptor to an internal handle
+    descriptor_table_get_ref(fildes, &entry);
+    if (entry->tag == DESCRIPTOR_TABLE_ENTRY_FILE_HANDLE) {
+      create_new_stream = true;
+
+      // Get the output stream
+      filesystem_own_output_stream_t stream_owned;
+      ok = filesystem_method_descriptor_write_via_stream(entry->file.file_handle,
+                                                         0,
+                                                         &stream_owned,
+                                                         &error_code);
+      if (!ok) {
+        translate_error(error_code);
+        return -1;
+      }
+      output_stream = streams_borrow_output_stream(stream_owned);
+    } else if (entry->tag == DESCRIPTOR_TABLE_ENTRY_FILE_STREAM) {
+      if (!entry->stream.file_info.writable) {
+        errno = EBADF;
+        return -1;
+      }
+      output_stream = entry->stream.write_stream;
+    } else {
+      errno = EBADF;
+      return -1;
+    }
+  }
+
+  // Check readiness for writing
+  uint64_t num_bytes_permitted = 0;
+  streams_stream_error_t stream_error;
+  ok = streams_method_output_stream_check_write(output_stream,
+                                                &num_bytes_permitted,
+                                                &stream_error);
+  if (!ok) {
+    errno = EIO;
+    return -1;
+  }
+
+  if (num_bytes_permitted < nbyte) {
+    streams_own_pollable_t pollable = streams_method_output_stream_subscribe(output_stream);
+    poll_method_pollable_block(poll_borrow_pollable(pollable));
```

Comment:
This pollable gets leaked - it needs a drop call. But also, can we cache it for next time this output stream is used? Creating a new pollable and dropping it each time creates more churn than ideal.

Reply:
Done in [45a0322](https://github.com/WebAssembly/wasi-libc/pull/606/commits/45a03227b38bd2b1218255323445a7cb58e49c86) -- but note that the pollables for `stdout` and `stderr` are not cached, since I'm not sure when they would be dropped if we were to cache them.

Reply:
I realized that this wasn't the best solution, as it creates pollables for any writable file (which may not be used if the file is only read from), so I also made pollables lazily initialized, in [e8dc233](https://github.com/WebAssembly/wasi-libc/pull/606/commits/e8dc233ff6ece3c73edbe003f8e3623487a69454).

I realize there will need to be a second pollable stored in the descriptor table, for the read stream, but I'll add that in #608 .